tracing/middleware.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022 CloudWeGo Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"context"
+	"time"
+
+	"github.com/bytedance/gopkg/cloud/metainfo"
+	"github.com/cloudwego/kitex/pkg/endpoint"
+	"github.com/cloudwego/kitex/pkg/klog"
+	"github.com/cloudwego/kitex/pkg/rpcinfo"
+	"github.com/kitex-contrib/obs-opentelemetry/tracing/internal"
+	"go.opentelemetry.io/otel/baggage"
+	"go.opentelemetry.io/otel/sdk/trace"
+	oteltrace "go.opentelemetry.io/otel/trace"
+)
+
+// ClientMiddleware inject span context into req meta
+func ClientMiddleware(cfg *config) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, req, resp interface{}) (err error) {
+			span := oteltrace.SpanFromContext(ctx)
+			if !span.IsRecording() {
+				return next(ctx, req, resp)
+			}
+
+			readOnlySpan := span.(trace.ReadOnlySpan)
+
+			// inject client service resource attributes (canonical service) to meta info
+			md := injectPeerServiceToMetaInfo(ctx, readOnlySpan.Resource().Attributes())
+
+			Inject(ctx, cfg, md)
+
+			for k, v := range md {
+				ctx = metainfo.WithValue(ctx, k, v)
+			}
+
+			return next(ctx, req, resp)
+		}
+	}
+}
+
+// ServerMiddleware extract req meta into span context
+func ServerMiddleware(cfg *config) endpoint.Middleware {
+	return func(next endpoint.Endpoint) endpoint.Endpoint {
+		return func(ctx context.Context, req, resp interface{}) (err error) {
+			tc := internal.TraceCarrierFromContext(ctx)
+			if tc == nil {
+				klog.CtxWarnf(ctx, "TraceCarrier not found in context")
+				return next(ctx, req, resp)
+			}
+
+			// get tracer from carrier
+			sTracer := tc.Tracer()
+
+			var spanName string
+			if cfg.spanNameFormatter != nil {
+				spanName = cfg.spanNameFormatter(ctx)
+			}
+
+			ri := rpcinfo.GetRPCInfo(ctx)
+			opts := []oteltrace.SpanStartOption{
+				oteltrace.WithTimestamp(getStartTimeOrDefault(ri, time.Now())),
```

Comment:
I think it is better to get time inside of getStartTimeOrDefault, it is an extra consumption here

Reply:
> I think it is better to get time inside of getStartTimeOrDefault, it is an extra consumption here

Makes sense, I'll change it